sympy/polys/matrices/rref.py

```diff
@@ -0,0 +1,393 @@
+    if K.is_EX and dense:
+        return _dm_rref_gj_div_dense(M)
+
+    # The sparse implementations are always faster when we can use them.
+    if dense:
+        M = M.to_sparse()
+
+    if K.is_QQ:
+        M_rref, pivots = _dm_rref_QQ(M)
+    else:
+        # This is definitely not optimal. More work needs to be done to
+        # determine the best algorithm for each domain. Generally it is better
+        # to use rref_den for nontrivial field domains in any case though.
+        M_rref, pivots = _dm_rref_gj_div(M)
+
+    if dense:
+        M_rref = M_rref.to_dense()
+
+    return M_rref, pivots
+
+
+def dm_rref_den(M, keep_domain=True):
+    """
+    Compute the reduced row echelon form of a matrix with denominator.
+
+    Examples
+    ========
+
+    >>> from sympy import ZZ
+    >>> from sympy.polys.matrices import DM
+    >>> from sympy.polys.matrices.rref import dm_rref_den
+    >>> M = DM([[1, 2], [3, 4]], ZZ)
+    >>> M_rref, den, pivots = dm_rref_den(M)
+    >>> M_rref.to_Matrix()
+    Matrix([
+    [-2,  0],
+    [ 0, -2]])
+    >>> den
+    -2
+    >>> pivots
+    (0, 1)
+
+    Parameters
+    ==========
+
+    keep_domain : bool, optional
+        If ``True`` then the domain of the matrix is preserved. If ``False``
+        then the domain might be preserved or it might be changed to an
+        associated field or ring (if that is potentially faster).
+
+    See Also
+    ========
+
+    sympy.polys.matrices.domainmatrix.DomainMatrix.rref_den
+        The ``DomainMatrix`` method that calls this function.
+    """
+    K = M.domain
+
+    # The sparse implementations are always faster
+    dense = M.rep.fmt == 'dense'
+    if dense:
+        M = M.to_sparse()
+
+    if K.is_ZZ:
+        M_rref, den, pivots = _dm_rref_den_ZZ(M, keep_domain=keep_domain)
+    elif K.is_QQ:
+        M_rref, den, pivots = _dm_rref_den_QQ(M, keep_domain=keep_domain)
+    else:
+        M_rref, den, pivots = _dm_rref_den_gj_ff(M)
+
+    if dense:
+        M_rref = M_rref.to_dense()
+
+    return M_rref, den, pivots
+
+
+def _dm_rref_gj_div(M):
+    """Compute RREF using sparse Gauss-Jordan elimination with division."""
+    M_rref_d, pivots, _ = sdm_irref(M.rep)
+    M_rref_sdm = SDM(M_rref_d, M.shape, M.domain)
+    pivots = tuple(pivots)
+    return M.from_rep(M_rref_sdm), pivots
+
+
+def _dm_rref_gj_div_dense(M):
+    """Compute RREF using dense Gauss-Jordan elimination with division."""
+    ddm = M.rep.copy()
+    pivots = ddm_irref(ddm)
+    M_rref_ddm = DDM(ddm, M.shape, M.domain)
+    pivots = tuple(pivots)
+    return M.from_rep(M_rref_ddm), pivots
+
+
+def _dm_rref_den_gj_ff(M):
+    """Compute RREF using sparse fraction-free Gauss-Jordan elimination."""
+    M_rref_d, den, pivots = sdm_rref_den(M.rep, M.domain)
+    M_rref_sdm = SDM(M_rref_d, M.shape, M.domain)
+    pivots = tuple(pivots)
+    return M.from_rep(M_rref_sdm), den, pivots
+
+
+def _dm_rref_den_ZZ(Mz, keep_domain=True):
```

Comment:
Could this, and each of the other functions that starts by selecting a method, accept a `method` kwarg, for users who wanted to choose the method themselves, instead of accepting the heuristic choice?

Reply:
I guess they could. I might need to think of better names for the methods though rather than `rref_QQ`. In future the same kinds of methods would be considered for different domains like `QQ(X)` etc.

Reply:
Perhaps for `rref_den` it should be:

1. `GJ`: Gauss Jordan with division. Always returns a result in the associated field.
2. `FFGJ`: Use fraction-free Gauss Jordan, always returns the same domain.
3. `CD`: If a field then clear denominators and convert to ring, always returns a result in the associated ring.

Those methods make sense for `rref_den`. For `rref` we don't have the option to return a result in a ring because generally the output needs to be in a field. Perhaps then it could be the same method names but with a slightly different meaning:

1. `GJ`: Convert to field and use Gauss Jordan with division. Always returns a result in the associated field.
2. `FFGJ`: Use fraction-free Gauss Jordan, convert to field and divide at the end, always returns a result in the associated field.
3. `CD`: If a field then clear denominators and convert to ring, use FFGJ but then convert to field and divide at the end, always returns a result in the associated field.

Maybe there are better names for these algorithms/approaches...

Reply:
I've added a method parameter e.g.:
```python
In [2]: M = Matrix([[1,1,1],[1,2,3]])

In [3]: M.to_DM().rref(method='GJ')
Out[3]: (DomainMatrix({0: {0: 1, 2: -1}, 1: {1: 1, 2: 2}}, (2, 3), QQ), (0, 1))
```
Possible methods are: auto, GJ, FF, CD, GJ_dense, FF_dense, CD_dense.

The parameter is for both rref and rref_den.

Reply:
> accept a method kwarg, for users who wanted to choose the method themselves, instead of accepting the heuristic choice?

My opinion is against that approach.
There is little of such usage to choose methods, and little usage of educating users to pick the methods,
unless the results are very different.

And it's just better to publicize more lower level handles, and let users program with them, than implementing many methods.

I've seen people who are looking for 'deep' features are actually looking for things like numpy C wrappers directly, rather than fiddling with 'methods'.

Reply:
I've added the method parameter to `DomainMatrix.rref` but not to `Matrix.rref`. There is a tension here between whether or not `DomainMatrix` methods should implement specific algorithms or should choose algorithms on behalf of the user. Until this PR every DomainMatrix method simply executes a specific algorithm. In the end though we will somewhere want to have something like an RREF method that can select the best algorithm if we know that different choices are better in different situations.

I can see though that `method=` arguments are awkward later because:

1. Most users don't use them.
2. A high-level operation like `solve` that calls e.g. `Matrix.rref` indirectly will not provide users any way to control the method anyway.
3. The method arguments make it awkward to change the internal details in future because they expose the internals too much.

Many Matrix methods have arguments like `iszerofunc` or `simplifyfunc`. This makes it awkward to e.g. change the method so that it uses `DomainMatrix` even if there is a situation where `DomainMatrix` is definitely better than any combination of existing options.

I think that the difference between these methods is not quite the same as the existing Matrix methods that have a `method=` argument because here the differences are really about what fundamental arithmetic operations a domain supports though. Also we already need to have the code that selects a method and then dispatches based on that regardless of whether we want to expose the method argument to users.